Run-time crash (undefined symbols) with OpenSSL 1.1.0-pre1 to 1.1.0-pre3

Output of `make test` with OpenSSL 1.1.0-pre1:

```
"/usr/bin/perl5.22.2" -MExtUtils::Command::MM -e 'cp_nonempty' -- SSLeay.bs blib/arch/auto/Net/SSLeay/SSLeay.bs 644
PERL_DL_NONLAZY=1 "/usr/bin/perl5.22.2" "-MExtUtils::Command::MM" "-MTest::Harness" "-e" "undef *Test::Harness::Switches; test_harness(0, 'blib/lib', 'blib/arch')" t/local/*.t t/handle/local/*.t
t/handle/local/05_use.t ................ 1/1
#   Failed test 'use Net::SSLeay::Handle;'
#   at t/handle/local/05_use.t line 8.
#     Tried to use 'Net::SSLeay::Handle'.
#     Error:  Can't load '/home/csn/contrib/p5-net-ssleay/blib/arch/auto/Net/SSLeay/SSLeay.so' for module Net::SSLeay: /home/csn/contrib/p5-net-ssleay/blib/arch/auto/Net/SSLeay/SSLeay.so: undefined symbol: OCSP_SINGLERESP_get0_id at /usr/lib64/perl5/DynaLoader.pm line 193.
#  at /home/csn/contrib/p5-net-ssleay/blib/lib/Net/SSLeay/Handle.pm line 8.
# Compilation failed in require at /home/csn/contrib/p5-net-ssleay/blib/lib/Net/SSLeay/Handle.pm line 8.
# BEGIN failed--compilation aborted at /home/csn/contrib/p5-net-ssleay/blib/lib/Net/SSLeay/Handle.pm line 8.
# Compilation failed in require at t/handle/local/05_use.t line 8.
# BEGIN failed--compilation aborted at t/handle/local/05_use.t line 8.
# Looks like you failed 1 test of 1.
t/handle/local/05_use.t ................ Dubious, test returned 1 (wstat 256, 0x100)
Failed 1/1 subtests
t/local/01_pod.t ....................... ok
t/local/02_pod_coverage.t .............. skipped: These tests are for only for release candidate testing. Enable with RELEASE_TESTING=1
t/local/03_use.t ....................... 1/?
#   Failed test 'use Net::SSLeay;'
#   at t/local/03_use.t line 11.
#     Tried to use 'Net::SSLeay'.
#     Error:  Can't load '/home/csn/contrib/p5-net-ssleay/blib/arch/auto/Net/SSLeay/SSLeay.so' for module Net::SSLeay: /home/csn/contrib/p5-net-ssleay/blib/arch/auto/Net/SSLeay/SSLeay.so: undefined symbol: OCSP_SINGLERESP_get0_id at /usr/lib64/perl5/DynaLoader.pm line 193.
#  at t/local/03_use.t line 11.
# Compilation failed in require at t/local/03_use.t line 11.
# BEGIN failed--compilation aborted at t/local/03_use.t line 11.
#
# Testing Net::SSLeay 1.89_02
#
# Perl information:
#   Version:         '5.022002'
#   Executable path: '/usr/bin/perl5.22.2'
#
# libssl information:
Makefile:1093: recipe for target 'test_dynamic' failed
make: *** [test_dynamic] Interrupt
```

[`OCSP_SINGLERESP_get0_id` wasn't added to OpenSSL until 1.0.0-pre3](https://github.com/openssl/openssl/commit/9e5cd4bac777e27ebcdc9aa411f0a63c27500468), so the version guards are likely incorrect. The output of `nm -D` on `SSLeay.so` indicates that `SSL_CTX_set_max_proto_version`, `SSL_CTX_set_min_proto_version`, `SSL_SESSION_up_ref`, `SSL_set_max_proto_version` and `SSL_set_min_proto_version` are also likely to be affected.

--- a/Changes
+++ b/Changes
@@ -1,5 +1,16 @@
 Revision history for Perl extension Net::SSLeay.
 
+1.89_03 ????
+	- Correct the minimum OpenSSL version required for the following functions
+	  to be made available (previously they were all declared to be present in
+	  1.1.0-pre1, which caused Net::SSLeay to crash at run-time when built
+	  against OpenSSL versions between 1.1.0-pre1 and 1.1.0-pre3):
+	  - CTX_set_max_proto_version (added in 1.1.0-pre2)
+	  - CTX_set_min_proto_version (added in 1.1.0-pre2)
+	  - SESSION_up_ref (added in 1.1.0-pre4)
+	  - set_max_proto_version (added in 1.1.0-pre2)
+	  - set_min_proto_version (added in 1.1.0-pre2)
+
 1.89_02 2020-08-07
 	- Add support for the P_X509_CRL_add_extensions function. Thanks to
 	  Manuel Mausz for the patch.

--- a/SSLeay.xs
+++ b/SSLeay.xs
@@ -2585,7 +2585,7 @@ d2i_SSL_SESSION(pv)
     OUTPUT:
 	RETVAL
 
-#if (OPENSSL_VERSION_NUMBER >= 0x10100000L && !defined(LIBRESSL_VERSION_NUMBER)) || (LIBRESSL_VERSION_NUMBER >= 0x2070000fL)
+#if (OPENSSL_VERSION_NUMBER >= 0x10100004L && !defined(LIBRESSL_VERSION_NUMBER)) || (LIBRESSL_VERSION_NUMBER >= 0x2070000fL)
 
 int
 SSL_SESSION_up_ref(sess)
@@ -4822,7 +4822,7 @@ TLS_client_method()
 #endif /* OpenSSL 1.1.0 or LibreSSL 2.2.2 */
 
 
-#if  (OPENSSL_VERSION_NUMBER >= 0x10100000L && !defined(LIBRESSL_VERSION_NUMBER)) || (LIBRESSL_VERSION_NUMBER >= 0x2060000fL)
+#if  (OPENSSL_VERSION_NUMBER >= 0x10100002L && !defined(LIBRESSL_VERSION_NUMBER)) || (LIBRESSL_VERSION_NUMBER >= 0x2060000fL)
 
 int
 SSL_CTX_set_min_proto_version(ctx, version)
@@ -4844,7 +4844,7 @@ SSL_set_max_proto_version(ssl, version)
      SSL *  ssl
      int    version
 
-#endif /* OpenSSL 1.1.0 or LibreSSL 2.6.0 */
+#endif /* OpenSSL 1.1.0-pre2 or LibreSSL 2.6.0 */
 
 
 #if OPENSSL_VERSION_NUMBER >= 0x1010007fL && !defined(LIBRESSL_VERSION_NUMBER)
@@ -7349,7 +7349,7 @@ OCSP_response_results(rsp,...)
 		if (!idsv) {
 		    /* getall: create new SV with OCSP_CERTID */
 		    unsigned char *pi,*pc;
-#if OPENSSL_VERSION_NUMBER >= 0x10100000L && !defined(LIBRESSL_VERSION_NUMBER)
+#if OPENSSL_VERSION_NUMBER >= 0x10100003L && !defined(LIBRESSL_VERSION_NUMBER)
 		    int len = i2d_OCSP_CERTID(OCSP_SINGLERESP_get0_id(sir),NULL);
 #else
 		    int len = i2d_OCSP_CERTID(sir->certId,NULL);
@@ -7358,7 +7358,7 @@ OCSP_response_results(rsp,...)
 		    Newx(pc,len,unsigned char);
 		    if (!pc) croak("out of memory");
 		    pi = pc;
-#if OPENSSL_VERSION_NUMBER >= 0x10100000L && !defined(LIBRESSL_VERSION_NUMBER)
+#if OPENSSL_VERSION_NUMBER >= 0x10100003L && !defined(LIBRESSL_VERSION_NUMBER)
 		    i2d_OCSP_CERTID(OCSP_SINGLERESP_get0_id(sir),&pi);
 #else
 		    i2d_OCSP_CERTID(sir->certId,&pi);

--- a/lib/Net/SSLeay.pod
+++ b/lib/Net/SSLeay.pod
@@ -1855,7 +1855,7 @@ Check openssl doc L<http://www.openssl.org/docs/ssl/SSL_SESSION_free.html|http:/
 
 =item * SESSION_up_ref
 
-B<COMPATIBILITY:> not available in Net-SSLeay-1.85 and before; requires at least OpenSSL 1.1.0 or LibreSSL 2.7.0
+B<COMPATIBILITY:> not available in Net-SSLeay-1.85 and before; requires at least OpenSSL 1.1.0-pre4 or LibreSSL 2.7.0
 
 Increases the reference counter on a SSL_SESSION structure.
 
@@ -2683,7 +2683,7 @@ Check openssl doc L<http://www.openssl.org/docs/ssl/SSL_CTX_new.html|http://www.
 
 =item * CTX_set_min_proto_version, CTX_set_max_proto_version, set_min_proto_version and set_max_proto_version,
 
-B<COMPATIBILITY:> not available in Net-SSLeay-1.82 and before; requires at least OpenSSL 1.1.0 or LibreSSL 2.6.0
+B<COMPATIBILITY:> not available in Net-SSLeay-1.82 and before; requires at least OpenSSL 1.1.0-pre2 or LibreSSL 2.6.0
 
 Set the minimum and maximum supported protocol for $ctx or $ssl.
 

